### ISSUE TITLE
Reallow functions on passthrough nodes

### ISSUE THREAD

**Issue by arvoelke** [nengo/nengo#805]
I've been stumbling into this problem recently:

```
import nengo
with nengo.Network() as model:
    a = nengo.Node(size_in=1)
    b = nengo.Node(size_in=1)
    nengo.Connection(a, b, synapse=None, function=lambda x: -x)
```

`Validation error when setting 'Connection.function_info': Cannot apply functions to passthrough nodes`

As a hack, this can be solved by adding the identity function to `a`, so that it is no longer a passthrough node:

```
output=lambda t, x: x
```

I realize that what I'm asking is in part trying to revert #428, which, from what I understand, was made in order to discourage biologically implausible models. But there are a class of functions that can be computed on passthrough nodes that _are_ biologically plausible, that cannot be efficiently implemented as a `transform`, such as reshaping and elementwise products.

Specifically, in my case, I've combined 200k neurons into a single ensemble, where disjoint sets represent subensembles from an `EnsembleArray`. In order to decode each subensemble, I need to do a bunch of dot products on subsets of these activities. While in theory this could be implemented as a single linear transform, it would be a sparse (block diagonal) 200k by 10k transform, which is clearly infeasible (2B elements). On the other hand, this can be implemented efficiently as a function on a node that does the appropriate sequence of linear numpy operations.

```
def f(a):
    a_view = a.view()
    a_view.shape = (num_subensembles, neurons_per_subensemble)
    return gains*np.dot(a_view, decoders.T)
```

The reason the node has to be passthrough is because functions cannot be used on connections coming from neurons, and so the activities have to be copied to a node.

To address this problem, we could do a number of things:
- Continue using the hack of adding identity functions to make nodes no longer passthrough
- Throw a warning instead of disallowing it completely
- Support passing a flag or configuration option to allow this on a one-by-one basis
- Support functions on neurons, perhaps with one of the above two caveats
- Create some more advanced Nengo objects that tap into linear operations like reshaping and elementwise products
- Address the "root" problems of #778 and #738, so that workarounds like this become unnecessary

**Comment by tcstewar** [comment]
> Continue using the hack of adding identity functions to make nodes no longer passthrough

I'd much prefer that sort of work-around.  It's explicit about what's happening, and it's what I was envisioning as a workaround when we put in that restriction

> Address the "root" problems of #778 and #738, so that workarounds like this become unnecessary

I think that's the right long-term solution.  So my vote would be to keep with the identity function approach until #778 and #738 get resolved....  :)

**Comment by arvoelke** [comment]
> I'd much prefer that sort of work-around. It's explicit about what's happening, and it's what I was envisioning as a workaround when we put in that restriction

In this case, maybe the error message should point to somewhere in the documentation that explains why this restriction is in place and how to circumvent it when needed.

This might be appropriate alongside the much-needed documentation on different connection types between the different combinations of pre and post objects.

**Comment by tbekolay** [comment]
Documentation as described in this issue is now available as of #1244. In particular, the section on the passthrough node function exception can be found [here](https://github.com/nengo/nengo/pull/1244/files#diff-ed414555199a61afe85a02d1bf9f1338R212).
